packages/dds/merge-tree/src/perspective.ts

```diff
@@ -0,0 +1,239 @@
+ */
+export class PerspectiveImpl implements Perspective {
+	/**
+	 * @param _mergeTree - The {@link MergeTree} to view.
+	 * @param _sequenceNumber - The latest sequence number to include in the view.
+	 * @param _clientId - The client ID to view from.
+	 * @param _localSequenceNumber - The latest local sequence number to include in the view.
+	 */
+	public constructor(
+		private readonly _mergeTree: MergeTree,
+		private readonly _sequenceNumber: number,
+		private readonly _clientId: number,
+		private readonly _refSeq: number = _sequenceNumber,
+		private readonly _localSequenceNumber?: number,
+	) {}
+
+	/**
+	 * Returns the immediately adjacent segment in the specified direction from this perspective.
+	 * There may actually be multiple segments between the given segment and the returned segment,
+	 * but they were either inserted after this perspective, or have been removed or moved before this perspective.
+	 *
+	 * @param segment - The segment to start from.
+	 * @param forward - The direction to search.
+	 * @returns the next segment in the specified direction, or the start or end of the tree if there is no next segment.
+	 */
+	public nextSegment(segment: ISegment, forward: boolean = true): ISegment {
+		let next: ISegment | undefined;
+		const action = (seg: ISegment) => {
+			if (
+				isSegmentPresent(
+					seg,
+					this._sequenceNumber,
+					this._clientId,
+					this._refSeq,
+					this._localSequenceNumber,
+				)
+			) {
+				next = seg;
+				return LeafAction.Exit;
+			}
+		};
+		(forward ? forwardExcursion : backwardExcursion)(segment, action);
+		return next ?? (forward ? this._mergeTree.endOfTree : this._mergeTree.startOfTree);
+	}
+
+	/**
+	 * @param segment - The segment to start from.
+	 * @returns the previous segment, or the start of the tree if there is no previous segment.
+	 * @remarks This is a convenient equivalent to calling `nextSegment(segment, false)`.
+	 */
+	public previousSegment(segment: ISegment): ISegment {
+		return this.nextSegment(segment, false);
+	}
+}
+
+/**
+ * @param seg - The segment to check.
+ * @param seq - The latest sequence number to consider.
+ * @param clientId - The ID of the client to consider.
+ * @param refSeq - The reference sequence number,
+ * aka the latest sequence number to consider for other clients' changes.
+ * @param localSeq - The latest local sequence number to consider.
+ * @returns true iff this segment was removed in the given perspective.
+ */
+export function wasRemovedBefore(
+	seg: ISegment,
+	seq: number,
+	clientId: number,
+	refSeq: number,
+	localSeq?: number,
+): boolean {
+	if (
+		seg.removedSeq === UnassignedSequenceNumber &&
+		localSeq !== undefined &&
+		seg.localRemovedSeq !== undefined
+	) {
+		return seg.localRemovedSeq <= localSeq;
+	}
+	if (seg.removedClientIds !== undefined && seg.removedClientIds.includes(clientId)) {
+		return (
+			seg.removedSeq !== undefined &&
+			seg.removedSeq !== UnassignedSequenceNumber &&
+			seg.removedSeq <= seq
```

Comment:
i'm a bit confused by seq, and how it differs from refseq? do we need it independently? i also think refSeq would be the correct value here

Reply:
it looks like it always matches refSeq, so i'd rather only keep that, to keep things conceptually consistent

Reply:
`seq` is the current sequence number. If this perspective is for an op, that's the seqNum assigned to that op. It may be different that refSeq because ops by the same clientId are considered up to `seq` instead of `refSeq`.

Reply:
Perspective and related functions no longer take `clientId` and `seq`, only `refSeq` and `localSeq`